### ISSUE TITLE
allow minimum rows to go to "1" and address dependency on current minimum of "5"

### ISSUE THREAD

**Issue by amanda-phet** [phetsims/plinko-probability#84]
One of our collaborators who observed sims in the classroom this summer had the suggestion to allow the board to go down to a single peg. I agree with him that it would really emphasize the concept of binary probability, and would allow teachers to more flexibly teach this concept (when they are usually teaching it with coins that don't have an adjustable probability). 

Is the code structured in a way that allows you to change the number of rows easily? I understand that a single row would probably scale the peg to be very large, but it seems worth it.

**Comment by pixelzoom** [comment]
I'm sorry to report that very bad things happen if I change the minimum number of rows to 1.   Some examples:

First, the size of the pegs and placement of their shadows is mess up.  Here's what the "Intro" screen looks like at startup ("Lab" screen is similar):

![screenshot_188](https://cloud.githubusercontent.com/assets/3046552/18528795/a94af374-7a87-11e6-984c-06b499b15717.png)

Here's what the "Lab" screen looks like when rows is set to 1. Both the peg and its shadow are not properly positioned, and (I believe) the peg should be larger:

![screenshot_189](https://cloud.githubusercontent.com/assets/3046552/18528850/e741c22a-7a87-11e6-955a-73124f8c91f3.png)

Next is the issue of ball size. With rows=1, here's what you get when the Lab screen is dispensing continuous balls:

![screenshot_195](https://cloud.githubusercontent.com/assets/3046552/18528987/822281a8-7a88-11e6-842f-09fb6f75e546.png)

I poked around, and there's no easy fix for any of these problems. And some of them (like setting a maximum ball size) would require a design meeting or two to work out.

@amanda-phet How do you want to proceed?

**Comment by pixelzoom** [comment]
If you want this change, I'd estimate somewhere in the neighborhood of 20-30 hours ferret out all the problem, decide how to address them, and adjust the implementation.  And we should stop dev testing immediately.

**Comment by pixelzoom** [comment]
Btw... The problems with the size and location of pegs and their shadows occurs for any number of rows < 5.  So there's something in the implementation (multiple somethings, actually) that are dependent on the value '5'.  This is a bad thing, but will be nontrivial to fix.

**Comment by pixelzoom** [comment]
Ugh. The problems with the size and location of pegs and their shadows occurs also occurs if the minimum number of rows > 5 !!  I'll add a note in the code so that someone doesn't step on this landmine.

**Comment by amanda-phet** [comment]
OK, that is a bummer... it is probably not worth 20-30 hours at this point! 

@ariel-phet @kathy-phet do you agree?

**Comment by ariel-phet** [comment]
@amanda-phet I don't think this would be appropriate at this time considering @pixelzoom's schedule 

However, having the code written in a way that depends on the minimum value of 5 is problematic and we should address. I think this would be something we could consider for Jesse once he is done Pendulum Lab (which is next for him after he finishes GAO). He appreciates "smallish" projects.

At worse case, we could have Martin and his team rework things next summer for a 2.0 release

**Comment by amanda-phet** [comment]
Sounds good to me. 

Unassigning.

**Comment by ariel-phet** [comment]
@pixelzoom thanks for the prompt investigation

**Comment by kathy-phet** [comment]
Thanks for investigating. And I agree - defer, but we should probably log a future issue too, to refine the code regardless. Jesse sounds like a good candidate at some point if it seems good to him.

**Comment by pixelzoom** [comment]
@kathy-phet said:
>  we should probably log a future issue too, to refine the code regardless.

This issue will suffice.

**Comment by kathy-phet** [comment]
ok

**Comment by pixelzoom** [comment]
9/15/16 dev meeting: Make a note in the code review checklist to monkey around with ranges during review, to check for this kind of issue.

**Comment by amanda-phet** [comment]
Quick ping- here is the issue that @pixelzoom estimated 20-30 hours to address. This sim will be the next priority the math grant.

**Comment by pixelzoom** [comment]
@ariel-phet @amanda-phet I reviewed this issue again, and must retract my statement earlier today that fixing this would probably be "a quickie".   Looks like it's going to be more involved, and I won't know exactly how much time until I'm well into fixing it.

**Comment by veillette** [comment]
I'll take a look at it.

**Comment by veillette** [comment]
Setting the minimum number of rows to 1, for screen intro


![image](https://cloud.githubusercontent.com/assets/7633344/25045922/9775c784-20fc-11e7-9bf7-75099b51c9b9.png)


whereas for the lab screen 
![image](https://cloud.githubusercontent.com/assets/7633344/25045953/ccf2c1fa-20fc-11e7-81e3-38dd0f23f5fc.png)

**Comment by veillette** [comment]
and for the record here is what it looks for one row

![image](https://cloud.githubusercontent.com/assets/7633344/25046009/1314f2a2-20fd-11e7-89f1-41d517467bb9.png)

**Comment by veillette** [comment]
if you set the minimum number of Rows to 1 and set the default value to 3, here is what the intro tab looks like:
![image](https://cloud.githubusercontent.com/assets/7633344/25046101/8d5c2ab2-20fd-11e7-8a99-1eeb5549fcfb.png)

**Comment by veillette** [comment]
The two previous commits only addressed the second part of this issue, namely the dependency on the number of rows being equal to five. I think you would need a design meeting to decide on the best scaling for the balls/ cylinders/pegs/hopper when the number of rows is small.

**Comment by veillette** [comment]
There are a lot of small details that become more prevalent when the number of rows get small.
(1) the peg shadow does not look realistic when the number of rows is equal to 1
(2) the diameter of the balls is larger than the hopper
(4) in the intro tab, the balls at the bottom of the cylinder get clipped
(5) the cylinders get filled up very quickly
(6) for n=1 the balls seems to go through the pegs.

**Comment by veillette** [comment]
The clipping of the balls in the cylinder is due to the canvasBounds of the balls that is determined by 
````
    var ballViewBounds = this.modelViewTransform.modelToViewBounds( ballModelBounds ).dilated( 20 );
```` 
the 20 should be increased and set to be proportional to the radius of the largest ball.

**Comment by ariel-phet** [comment]
@veillette it seems that the scaling shown in https://github.com/phetsims/plinko-probability/issues/84#issuecomment-294163728 looks good. Could we try a rule the for rows being 3 or less the scaling is fixed at what you have shown?

**Comment by veillette** [comment]
I added query parameters minRow, maxRow and defaultRow so you can see for yourself. 
eg. plinko-probability_en.html?minRow=1&defaultRow=3
will set the minimum number of Row to 1

Below are some screenshots. FYI, I only changed the size of the balls, which is relatively trivial to do.
Changing the size of the balls in such arbitrary manner affects the intro screen if the default value is set below three (by default it is 12 so it is not affected by the change). You will see that the balls do not stack neatly in the cylinder bins. 
 
Changing the size of the pegs is easy but the balls do not bounce off the pegs anymore since the pegs and the balls are no longer bound by the proper scaling. In other words, it would take a bit more effort to change the size of the pegs.

![image](https://cloud.githubusercontent.com/assets/7633344/25066311/5e87b762-21f0-11e7-8b15-428d3f295e17.png)

![image](https://cloud.githubusercontent.com/assets/7633344/25066317/8ed7cfba-21f0-11e7-88f0-d3218eed2378.png)

**Comment by veillette** [comment]
If you are serious about setting n=1 and having worked nicely so that the balls seem to bounced off the pegs (rather than just arbitrarily setting the radius of the balls and the pegs) , it would be preferable to work a branch rather than master.

**Comment by ariel-phet** [comment]
@veillette yes, we are serious about that

**Comment by pixelzoom** [comment]
[4/19/17, 11:41:24 AM] Ariel Paul: Hey Chris, Martin said he is happy to do the Plinko work, and has some time over the next few weeks to work away at it

Assigning to @veillette.

**Comment by veillette** [comment]
I addressed the issue of scaling the pegs and the balls for the number of rows 1 to 4 in #99.

Closing

**Comment by samreid** [comment]
There are still TODOs marked for this issue, discovered during https://github.com/phetsims/chipper/issues/946

**Comment by samreid** [comment]
Unassigning until we return focus to this sim.
